ext/statsd/statsd.c

```diff
@@ -0,0 +1,376 @@
+normalized_names_cached(struct datagram_builder *builder, VALUE self, VALUE name)
+{
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  st_index_t key;
+  st_data_t val;
+  VALUE cached;
+  key = (st_index_t)RSTRING_PTR(name);
+  if (st_lookup(builder->normalized_names_cache, key, &val)){
+    return (VALUE)val;
+  } else if (builder->normalized_names_cache->num_entries < NORMALIZED_NAMES_CACHE_MAX) {
+    cached = normalize_name(builder, self, name);
+    st_insert(builder->normalized_names_cache, key, (st_data_t)cached);
+    return cached;
+  }
+  return normalize_name(builder, self, name);
+  RB_GC_GUARD(cached);
+#else
+  return normalize_name(self, name);
+#endif
+}
+
+/* pure function not exposed to ruby with an intermediate bounded cache */
+static VALUE
+normalized_tags_cached(struct datagram_builder *builder, VALUE self, VALUE tags)
+{
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_index_t key;
+  st_data_t val;
+  VALUE cached;
+  // More involved hashing as we need to hash on the content of the container too
+  // XXX: revisit
+  key = (st_index_t)(FIX2LONG(rb_hash(tags)));
+  if (st_lookup(builder->normalized_tags_cache, key, &val)){
+    return (VALUE)val;
+  } else if (builder->normalized_tags_cache->num_entries < NORMALIZED_TAGS_CACHE_MAX) {
+    cached = rb_funcall(self, idNormalizeTags, 1, tags);
+    st_insert(builder->normalized_tags_cache, key, (st_data_t)cached);
+    return cached;
+  }
+  return rb_funcall(self, idNormalizeTags, 1, tags);
+  RB_GC_GUARD(cached);
+#else
+  return rb_funcall(self, idNormalizeTags, 1, tags);
+#endif
+}
+
+inline static bool append_normalized_tags(struct datagram_builder *builder, VALUE normalized_tags, int trim_trailing_comma)
+{
+  VALUE tag;
+  int tags_len = 0, chunk_len = 0, i = 0;
+  tags_len = (int)RARRAY_LEN(normalized_tags);
+  for (i = 0; i < tags_len; ++i) {
+    tag = RARRAY_AREF(normalized_tags, i);
+    chunk_len = (int)RSTRING_LEN(tag);
+    if (builder->len + chunk_len > DATAGRAM_SIZE_MAX) return false;
+    memcpy(builder->datagram + builder->len, StringValuePtr(tag), chunk_len);
+    builder->len += chunk_len;
+    if (!trim_trailing_comma || i < tags_len - 1) {
+      if (builder->len + 1 > DATAGRAM_SIZE_MAX) return false;
+      memcpy(builder->datagram + builder->len, ",", 1);
+      builder->len += 1;
+    }
+  }
+  return true;
+}
+
+static VALUE
+generate_generic_datagram(VALUE self, VALUE name, VALUE value, const char *type, VALUE sample_rate, VALUE tags) {
+  VALUE normalized_name, str_value, str_sample_rate;
+  VALUE normalized_tags = Qnil;
+  char sr_buf[SAMPLE_RATE_SIZE_MAX];
+  bool empty_tags = true;
+  long chunk_len = 0;
+  get_datagram_builder_struct(self);
+
+  builder->len = builder->prefix_len;
+
+  if (NIL_P(normalized_name = normalize_name_fast_path(self, name))) {
+    normalized_name = normalized_names_cached(builder, self, name);
+  }
+  chunk_len = RSTRING_LEN(normalized_name);
+  if (builder->len + chunk_len > DATAGRAM_SIZE_MAX) goto finalize_datagram;
```

Comment:
Should we be raising on these conditions?

Reply:
I thought about it, but then again this is telemetry specific and the default buffer is a generous buffer size of 4096 bytes per datagram.

The line is somewhere between:

* Raising on a telemetry specific path the could bubble up an unhandled exception, impacting business specific paths. I don't think that's a great default.

* If you emit datagrams that large chances are cardinality is also going to be high through tags most likely and a truncated datagram would be the least we could do.

Reply:
What's the behaviour if we send a datagram that's too large? Does that depend on the statsd backend?

Reply:
Currently, we don't know. I assume they will simply not be delivered because they are being dropped by the networking stack as the UDP datagrams get too large. By design, we don't know about this - it's a fire and forget protocol.

Reply:
What Willem said re. `fire and forget`. I'm fine with any solution other than raising an error as telemetry specific fire and forget datagrams shouldn't risk raising unhandled exceptions that could uproot business processes.